src/codec.rs

```diff
@@ -1113,71 +1067,110 @@ impl<T: Encode> Encode for [T] {
 	}
 }
 
-/// Create a `Vec<T>` by casting directly from a buffer of read `u8`s
-///
-/// The encoding of `T` must be equal to its binary representation, and size of `T` must be less or
-/// equal to [`MAX_PREALLOCATION`].
-pub(crate) fn read_vec_from_u8s<I, T>(input: &mut I, items_len: usize) -> Result<Vec<T>, Error>
+fn decode_vec_chunked<T, F>(len: usize, mut decode_chunk: F) -> Result<Vec<T>, Error>
 where
-	I: Input,
-	T: ToMutByteSlice + Default + Clone,
+	F: FnMut(&mut Vec<T>, usize) -> Result<(), Error>,
 {
 	debug_assert!(MAX_PREALLOCATION >= mem::size_of::<T>(), "Invalid precondition");
+	let chunk_len = MAX_PREALLOCATION / mem::size_of::<T>();
 
-	let byte_len = items_len
-		.checked_mul(mem::size_of::<T>())
-		.ok_or("Item is too big and cannot be allocated")?;
+	let mut decoded_vec = vec![];
+	let mut num_undecoded_items = len;
+	while num_undecoded_items > 0 {
+		let chunk_len = chunk_len.min(num_undecoded_items);
+		decoded_vec.reserve_exact(chunk_len);
 
-	let input_len = input.remaining_len()?;
+		decode_chunk(&mut decoded_vec, chunk_len)?;
 
-	// If there is input len and it cannot be pre-allocated then return directly.
-	if input_len.map(|l| l < byte_len).unwrap_or(false) {
-		return Err("Not enough data to decode vector".into());
+		num_undecoded_items = num_undecoded_items.saturating_sub(chunk_len);
```

Comment:
This `saturating_sub`'s completely unnecessary here since impossible to have `chunk_len > num_undecoded_items` due to the `min`.

Reply:
Addressed in #615